### PR TITLE
Fix mongo failed to parse `absolute-datetime`

### LINKED ISSUE #16828: Could not convert UTC to a ZoneOffset
**Logs**

```
ERROR middleware.catch-exceptions :: Error processing query: null
{:database_id 2,
 :started_at #t "2021-06-30T14:23:36.779561+08:00[Asia/Shanghai]",
 :json_query
 {:database 2,
  :query
  {:source-table 12,
   :aggregation [["count"]],
   :breakout [["field" 47 {:temporal-unit "week"}]],
   :filter ["between" ["field" 47 {:temporal-unit "month"}] "2020-12-01T00:00:00+08:00" "2021-02-01T00:00:00+08:00"]},
  :type "query",
  :parameters [],
  :middleware {:js-int-to-string? true, :add-default-userland-constraints? true}},
 :native nil,
 :status :failed,
 :class java.time.DateTimeException,
 :stacktrace
 ["java_time.zone$zone_offset.invokeStatic(zone.clj:69)"
  "java_time.zone$zone_offset.invoke(zone.clj:43)"
  "java_time.zone$fn__26209.invokeStatic(zone.clj:343)"
  "java_time.zone$fn__26209.invoke(zone.clj:334)"
  "java_time.zone$fn__26192$G__26174__26199.invoke(zone.clj:317)"
  "--> driver.mongo.query_processor$eval543$fn__545.invoke(query_processor.clj:245)"
  "driver.mongo.query_processor$filter_expr.invokeStatic(query_processor.clj:335)"
  "driver.mongo.query_processor$filter_expr.invoke(query_processor.clj:333)"
  "driver.mongo.query_processor$eval661$fn__663.invoke(query_processor.clj:356)"
  "driver.mongo.query_processor$eval669$fn__671.invoke(query_processor.clj:360)"
  "driver.mongo.query_processor$handle_filter.invokeStatic(query_processor.clj:388)"
  "driver.mongo.query_processor$handle_filter.invoke(query_processor.clj:385)"
  "driver.mongo.query_processor$eval1124$generate_aggregation_pipeline__1129$fn__1130$fn__1131.invoke(query_processor.clj:652)"
  "driver.mongo.query_processor$eval1124$generate_aggregation_pipeline__1129$fn__1130.invoke(query_processor.clj:651)"
  "driver.mongo.query_processor$eval1124$generate_aggregation_pipeline__1129.invoke(query_processor.clj:648)"
  "driver.mongo.query_processor$mbql__GT_native.invokeStatic(query_processor.clj:683)"
  "driver.mongo.query_processor$mbql__GT_native.invoke(query_processor.clj:676)"
  "driver.mongo$eval2415$fn__2416.invoke(mongo.clj:220)"
  "query_processor.middleware.mbql_to_native$query__GT_native_form.invokeStatic(mbql_to_native.clj:14)"
  "query_processor.middleware.mbql_to_native$query__GT_native_form.invoke(mbql_to_native.clj:9)"
  "query_processor.middleware.mbql_to_native$mbql__GT_native$fn__47664.invoke(mbql_to_native.clj:22)"
  "query_processor.middleware.check_features$check_features$fn__46778.invoke(check_features.clj:39)"
  "query_processor.middleware.limit$limit$fn__47650.invoke(limit.clj:37)"
  "query_processor.middleware.cache$maybe_return_cached_results$fn__46230.invoke(cache.clj:211)"
  "query_processor.middleware.optimize_temporal_filters$optimize_temporal_filters$fn__47910.invoke(optimize_temporal_filters.clj:204)"
  "query_processor.middleware.validate_temporal_bucketing$validate_temporal_bucketing$fn__49840.invoke(validate_temporal_bucketing.clj:50)"
  "query_processor.middleware.auto_parse_filter_values$auto_parse_filter_values$fn__45349.invoke(auto_parse_filter_values.clj:43)"
  "query_processor.middleware.wrap_value_literals$wrap_value_literals$fn__41497.invoke(wrap_value_literals.clj:161)"
  "query_processor.middleware.annotate$add_column_info$fn__41372.invoke(annotate.clj:605)"
  "query_processor.middleware.permissions$check_query_permissions$fn__46650.invoke(permissions.clj:81)"
  "query_processor.middleware.pre_alias_aggregations$pre_alias_aggregations$fn__48768.invoke(pre_alias_aggregations.clj:40)"
  "query_processor.middleware.cumulative_aggregations$handle_cumulative_aggregations$fn__46851.invoke(cumulative_aggregations.clj:60)"
  "query_processor.middleware.resolve_joined_fields$resolve_joined_fields$fn__49067.invoke(resolve_joined_fields.clj:102)"
  "query_processor.middleware.resolve_joins$resolve_joins$fn__49380.invoke(resolve_joins.clj:171)"
  "query_processor.middleware.add_implicit_joins$add_implicit_joins$fn__44925.invoke(add_implicit_joins.clj:190)"
  "query_processor.middleware.large_int_id$convert_id_to_string$fn__47614.invoke(large_int_id.clj:59)"
  "query_processor.middleware.format_rows$format_rows$fn__47595.invoke(format_rows.clj:74)"
  "query_processor.middleware.add_default_temporal_unit$add_default_temporal_unit$fn__44219.invoke(add_default_temporal_unit.clj:23)"
  "query_processor.middleware.desugar$desugar$fn__46917.invoke(desugar.clj:21)"
  "query_processor.middleware.binning$update_binning_strategy$fn__45736.invoke(binning.clj:227)"
  "query_processor.middleware.resolve_fields$resolve_fields$fn__46453.invoke(resolve_fields.clj:34)"
  "query_processor.middleware.add_dimension_projections$add_remapping$fn__44574.invoke(add_dimension_projections.clj:314)"
  "query_processor.middleware.add_implicit_clauses$add_implicit_clauses$fn__44803.invoke(add_implicit_cla
uses.clj:147)"
  "query_processor.middleware.upgrade_field_literals$upgrade_field_literals$fn__49789.invoke(upgrade_field_literals.clj:40)"
  "query_processor.middleware.add_source_metadata$add_source_metadata_for_source_queries$fn__45088.invoke(add_source_metadata.clj:123)"
  "query_processor.middleware.reconcile_breakout_and_order_by_bucketing$reconcile_breakout_and_order_by_bucketing$fn__48942.invoke(reconcile_breakout_and_order_by_bucketing.clj:100)"
  "query_processor.middleware.auto_bucket_datetimes$auto_bucket_datetimes$fn__45296.invoke(auto_bucket_datetimes.clj:147)"
  "query_processor.middleware.resolve_source_table$resolve_source_tables$fn__46500.invoke(resolve_source_table.clj:45)"
  "query_processor.middleware.parameters$substitute_parameters$fn__48750.invoke(parameters.clj:111)"
  "query_processor.middleware.resolve_referenced$resolve_referenced_card_resources$fn__46552.invoke(resolve_referenced.clj:79)"
  "query_processor.middleware.expand_macros$expand_macros$fn__47301.invoke(expand_macros.clj:184)"
  "query_processor.middleware.add_timezone_info$add_timezone_info$fn__45097.invoke(add_timezone_info.clj:15)"
  "query_processor.middleware.splice_params_in_response$splice_params_in_response$fn__49742.invoke(splice_params_in_response.clj:32)"
  "query_processor.middleware.resolve_database_and_driver$resolve_database_and_driver$fn__48953$fn__48957.invoke(resolve_database_and_driver.clj:31)"
  "driver$do_with_driver.invokeStatic(driver.clj:60)"
  "driver$do_with_driver.invoke(driver.clj:56)"
  "query_processor.middleware.resolve_database_and_driver$resolve_database_and_driver$fn__48953.invoke(resolve_database_and_driver.clj:25)"
  "query_processor.middleware.fetch_source_query$resolve_card_id_source_tables$fn__47541.invoke(fetch_source_query.clj:274)"
  "query_processor.middleware.store$initialize_store$fn__49751$fn__49752.invoke(store.clj:11)"
  "query_processor.store$do_with_store.invokeStatic(store.clj:44)"
  "query_processor.store$do_with_store.invoke(store.clj:38)"
  "query_processor.middleware.store$initialize_store$fn__49751.invoke(store.clj:10)"
  "query_processor.middleware.validate$validate_query$fn__49796.invoke(validate.clj:10)"
  "query_processor.middleware.normalize_query$normalize$fn__47677.invoke(normalize_query.clj:22)"
  "query_processor.middleware.add_rows_truncated$add_rows_truncated$fn__44943.invoke(add_rows_truncated.clj:35)"
  "query_processor.middleware.results_metadata$record_and_return_metadata_BANG_$fn__49727.invoke(results_metadata.clj:147)"
  "query_processor.middleware.constraints$add_default_userland_constraints$fn__46794.invoke(constraints.clj:42)"
  "query_processor.middleware.process_userland_query$process_userland_query$fn__48839.invoke(process_userland_query.clj:135)"
  "query_processor.middleware.catch_exceptions$catch_exceptions$fn__46734.invoke(catch_exceptions.clj:173)"
  "query_processor.reducible$async_qp$qp_STAR___38051$thunk__38052.invoke(reducible.clj:103)"
  "query_processor.reducible$async_qp$qp_STAR___38051.invoke(reducible.clj:109)"
  "query_processor.reducible$sync_qp$qp_STAR___38060$fn__38063.invoke(reducible.clj:135)"
  "query_processor.reducible$sync_qp$qp_STAR___38060.invoke(reducible.clj:134)"
  "query_processor$process_userland_query.invokeStatic(query_processor.clj:241)"
  "query_processor$process_userland_query.doInvoke(query_processor.clj:237)"
  "query_processor$fn__49886$process_query_and_save_execution_BANG___49895$fn__49898.invoke(query_processor.clj:253)"
  "query_processor$fn__49886$process_query_and_save_execution_BANG___49895.invoke(query_processor.clj:245)"
  "query_processor$fn__49930$process_query_and_save_with_max_results_constraints_BANG___49939$fn__49942.invoke(query_processor.clj:265)"
  "query_processor$fn__49930$process_query_and_save_with_max_results_constraints_BANG___49939.invoke(query_processor.clj:258)"
  "api.dataset$run_query_async$fn__56143.invoke(dataset.clj:56)"
  "query_processor.streaming$streaming_response_STAR_$fn__56122$fn__56123.invoke(streaming.clj:72)"
  "query_processor.streaming$streaming_response_STAR_$fn__56122.invoke(streaming.clj:71)"
  "async.streaming_response$do_f_STAR_.invokeStatic(streaming_response.clj:65)"
  "async.streaming_response$do_f_STAR_.invoke(streaming_response.clj:63)"
  "async.streaming_response$do_f_async$fn__16071.invoke(streaming_response.clj:84)"],
 :context :ad-hoc,
 :error "Could not convert UTC to a ZoneOffset!",
 :row_count 0,
 :running_time 0,
 :preprocessed
 {:database 2,
  :query
  {:source-table 12,
   :aggregation [[:aggregation-options [:count] {:name "count"}]],
   :breakout [[:field 47 {:temporal-unit :week}]],
   :filter
   [:and
    [:>= [:field 47 {:temporal-unit :default}] [:absolute-datetime #t "2020-12-01T00:00+08:00" :default]]
    [:< [:field 47 {:temporal-unit :default}] [:absolute-datetime #t "2021-03-01T00:00+08:00" :default]]],
   :order-by [[:asc [:field 47 {:temporal-unit :week}]]]},
  :type :query,
  :middleware {:js-int-to-string? true, :add-default-userland-constraints? true},
  :info
  {:executed-by 1,
   :context :ad-hoc,
   :nested? false,
   :query-hash
   [10, 89, -68, -56, -117, -101, 118, -70, 32, -64, -75, 18, 41, 48, -106, 14, 3, -100, 68, 106, -43, 83, 19, -37, 54,
    60, -8, 80, -37, -98, 88, 75]},
  :constraints {:max-results 10000, :max-results-bare-rows 2000}},
 :data {:rows [], :cols []}}

2021-06-30 14:23:37,010 DEBUG middleware.log :: POST /api/dataset 202 [ASYNC: completed] 241.5 ms (8 DB calls) App DB connections: 0/4 Jetty threads: 2/50 (5 idle, 0 queued) (54 total active threads) Queries in flight: 0 (0 queued)
2021-06-30 14:24:25,322 DEBUG middleware.log :: POST /api/dataset 202 [ASYNC: completed] 132.1 ms (11 DB calls) App DB connections: 1/4 Jetty threads: 2/50 (5 idle, 0 queued) (55 total active threads) Queries in flight: 0 (0 queued)
2021-06-30 14:24:27,225 DEBUG middleware.log :: GET /api/card/1 200 16.7 ms (6 DB calls) App DB connections: 0/4 Jetty threads: 3/50 (4 idle, 0 queued) (55 total active threads) Queries in flight: 0 (0 queued)
2021-06-30 14:24:27,532 ERROR middleware.catch-exceptions :: Error processing query: null
{:database_id 2,
...skipping...
  "query_processor$process_userland_query.invokeStatic(query_processor.clj:241)"
  "query_processor$process_userland_query.doInvoke(query_processor.clj:237)"
  "query_processor$fn__49886$process_query_and_save_execution_BANG___49895$fn__49898.invoke(query_processor.clj:253)"
  "query_processor$fn__49886$process_query_and_save_execution_BANG___49895.invoke(query_processor.clj:245)"
  "query_processor$fn__49930$process_query_and_save_with_max_results_constraints_BANG___49939$fn__49942.invoke(query_processor.clj:265)"
  "query_processor$fn__49930$process_query_and_save_with_max_results_constraints_BANG___49939.invoke(query_processor.clj:258)"
  "api.dataset$run_query_async$fn__56143.invoke(dataset.clj:56)"
  "query_processor.streaming$streaming_response_STAR_$fn__56122$fn__56123.invoke(streaming.clj:72)"
  "query_processor.streaming$streaming_response_STAR_$fn__56122.invoke(streaming.clj:71)"
  "async.streaming_response$do_f_STAR_.invokeStatic(streaming_response.clj:65)"
  "async.streaming_response$do_f_STAR_.invoke(streaming_response.clj:63)"
  "async.streaming_response$do_f_async$fn__16071.invoke(streaming_response.clj:84)"],
 :context :ad-hoc,
 :error "Could not convert UTC to a ZoneOffset!",
 :row_count 0,
 :running_time 0,
 :preprocessed
 {:type :query,
  :query
  {:source-table 12,
   :aggregation [[:aggregation-options [:count] {:name "count"}]],
   :filter
   [:and
    [:>= [:field 47 {:temporal-unit :default}] [:absolute-datetime #t "2021-03-01T00:00+08:00" :default]]
    [:< [:field 47 {:temporal-unit :default}] [:absolute-datetime #t "2021-04-01T00:00+08:00" :default]]],
   :breakout [[:field 47 {:temporal-unit :week}]],
   :order-by [[:asc [:field 47 {:temporal-unit :week}]]]},
  :database 2,
  :middleware {:js-int-to-string? true, :add-default-userland-constraints? true},
  :info
  {:executed-by 1,
   :context :ad-hoc,
   :nested? false,
   :query-hash
   [118, -26, 71, 7, -111, 99, 96, 35, 2, -114, 118, 66, -93, -41, 55, 47, -90, -47, -38, 123, 106, -110, -18, 30, -67,
    61, 99, -96, -65, 119, 54, -104]},
  :constraints {:max-results 10000, :max-results-bare-rows 2000}},
 :data {:rows [], :cols []}}
```

**To Reproduce**
1. `Zoom in` a question that grouped by timestamp column.

**Screenshots**
image
image

**Information about Metabase Installation:**
```
metabase docker image: v0.39.4
application database: postgres v13.3

MB_REPORT_TIMEZONE: Asia/Shanghai
JAVA_TIMEZONE: Asia/Shanghai

connecting database: mongo v4.0
table timestamp column: "createdAt" : ISODate("2020-02-18T14:52:28.654Z") 
```

:arrow_down: Please click the :+1: reaction instead of leaving a `+1` or `update?` comment

### PR DESCRIPTION
**Pull request by qnkhuat**
Fixes #16828

The bug was because FE sent an absolute datetime with an `offset` (`2020-12-01T00:00:00+08:00`) when you zoom in on chart.

Even though it's a bit weird that FE will send a datetime `without offset` when you use the datetime filter. But we still need to fix our driver to handle all cases.

### CODE REVIEW

File: modules/drivers/mongo/src/metabase/driver/mongo/query_processor.clj [hunk 1]
```diff
@@ -325,12 +325,12 @@
...
-                      java.time.LocalTime      t
-                      java.time.LocalDateTime  t
-                      java.time.OffsetTime     (t/with-offset-same-instant t report-zone)
-                      java.time.OffsetDateTime (t/with-offset-same-instant t report-zone)
-                      java.time.ZonedDateTime  (t/offset-date-time (t/with-zone-same-instant t report-zone)))]
+                     java.time.LocalDate      t
+                     java.time.LocalTime      t
+                     java.time.LocalDateTime  t
+                     java.time.OffsetTime     (t/with-offset-same-instant t (.. report-zone getRules (getOffset (t/instant t))))
+                     java.time.OffsetDateTime (t/with-offset-same-instant t (.. report-zone getRules (getOffset (t/instant t))))
```
Comment (calherries) [inline modules/drivers/mongo/src/metabase/driver/mongo/query_processor.clj:332]: This is simpler
```
(t/offset-date-time (t/instant t) report-zone)
```

File: modules/drivers/mongo/src/metabase/driver/mongo/query_processor.clj [hunk 2]
```diff
@@ -325,12 +325,12 @@
...
+                     java.time.ZonedDateTime  (t/offset-date-time (t/with-zone-same-instant t report-zone)))]
```
Comment (calherries) [inline modules/drivers/mongo/src/metabase/driver/mongo/query_processor.clj:333]: This could also be changed to `(t/offset-date-time t report-zone)`

**Review by calherries — APPROVED** [review]
does the job 👍

**Comment by flamber** [comment]
@qnkhuat Is this going into 45 or 46?

**Comment by qnkhuat** [comment]
45 too!

**Pull request merged**